File: 算法/代码随想录算法PDF/00调试源代码/202.快乐数.py
#
# @lc app=leetcode.cn id=202 lang=python3
#
# [202] 快乐数
#

# @lc code=start
class Solution:
    def getHappy(self, n: int) -> int:
        res = 0
        while n:
            res += (n % 10) * (n % 10)
            n //= 10
        return res
    def isHappy(self, n: int) -> bool:
        Cycle = {4, 16, 37, 58, 89, 145, 42, 20}
        while n != 1 and n not in Cycle:
            n = self.getHappy(n)
        return n == 1
# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/202.快乐数.cpp
/*
 * @lc app=leetcode.cn id=202 lang=cpp
 *
 * [202] 快乐数
 */

// @lc code=start
class Solution {
public:
    int getHappy(int n) {  // 获取下一个happy数
        int res = 0;
        while (n) {
            res += (n % 10) * (n % 10);
            n /= 10;
        }
        return res;
    }

    bool isHappy(int n) {
        unordered_set<int> Cycle({4, 16, 37, 58, 89, 145, 42, 20});
        do {
            n = getHappy(n);
        } while (n != 1 && Cycle.find(n) == Cycle.end());
        return n == 1;
    }
};
// @lc code=end

